### PR TITLE
Np 711 retry getting when failure

### PR DESCRIPTION
**Pull request by axthosarouris**
Error to fix: When we receive a GOAWAY frame from CRISTIN or other servers, HttpClient throws an ExecutionException and  we send an error back to our REST-API client.

Fix: Wait for 500 msec, and then retry to get the data. If the HttpClient throws an Exception, then we try again and if HttpClient throws an Exception again then we send an error message back to our REST-API clients.

Changes:
1. upgrade nva-commons.
2. Update HttpExecutorImpl to retry in case of an ExectutionException.
3. Log the error as a warning.

### CODE REVIEW

File: src/main/java/no/unit/nva/institution/proxy/HttpExecutorImpl.java [hunk 1]
```diff
@@ -91,6 +91,54 @@ public JsonNode getNestedInstitution(URI uri, Language language) throws HttpClie
...
+    private Try<HttpResponse<String>> sendRequestMultipleTimes(URI uri) {
+        return tryGettingResponse(uri, FIRST_EFFORT, null);
+    }
+
+    private Try<HttpResponse<String>> tryGettingResponse(URI uri, int effortCount,
+                                                         Try<HttpResponse<String>> lastEffort) {
+        if (lastEffortWasASuccessOrWeFailedTooManyTimes(effortCount, lastEffort)) {
+            return lastEffort;
+        } else {
+            if (effortCount > 0) {
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/institution/proxy/HttpExecutorImpl.java]: FIRST_EFFORT?

File: src/main/java/no/unit/nva/institution/proxy/HttpExecutorImpl.java [hunk 2]
```diff
@@ -91,6 +91,54 @@ public JsonNode getNestedInstitution(URI uri, Language language) throws HttpClie
         return generator.getNestedInstitution();
     }
 
+    private Try<HttpResponse<String>> sendRequestMultipleTimes(URI uri) {
+        return tryGettingResponse(uri, FIRST_EFFORT, null);
+    }
+
+    private Try<HttpResponse<String>> tryGettingResponse(URI uri, int effortCount,
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/institution/proxy/HttpExecutorImpl.java]: This feels like it should be a loop

File: src/main/java/no/unit/nva/institution/proxy/HttpExecutorImpl.java [hunk 3]
```diff
@@ -91,6 +91,54 @@ public JsonNode getNestedInstitution(URI uri, Language language) throws HttpClie
...
+            Try<HttpResponse<String>> newEffort = attempt(() -> sendHttpRequest(uri).get());
+            if (newEffort.isFailure()) {
+                logger.warn("Failed HttpRequest:" + newEffort.getException().getMessage(), newEffort.getException());
+            }
+
+            return tryGettingResponse(uri, effortCount + 1, newEffort);
+        }
+    }
+
+    private CompletableFuture<HttpResponse<String>> sendHttpRequest(URI uri) {
```
Comment (janbj) [inline src/main/java/no/unit/nva/institution/proxy/HttpExecutorImpl.java]: createAndSendHttpRequest() ?

**Review by janbj — DISMISSED** [review]

**Review by janbj — DISMISSED** [review]

File: src/main/java/no/unit/nva/institution/proxy/HttpExecutorImpl.java [hunk 4]
```diff
@@ -91,6 +91,53 @@ public JsonNode getNestedInstitution(URI uri, Language language) throws HttpClie
         return generator.getNestedInstitution();
     }
 
+    private Try<HttpResponse<String>> sendRequestMultipleTimes(URI uri) {
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/institution/proxy/HttpExecutorImpl.java:94]:
```suggestion
    private Try<HttpResponse<String>> sendRequestMultipleTimes(URI uri) {
        AtomicReference<Try<HttpResponse<String>>> last = new AtomicReference<>();
        return IntStream.range(FIRST_EFFORT, MAX_EFFORTS)
                .filter(effort -> effort == FIRST_EFFORT || last.get().isFailure())
                .map(this::waitBeforeRetrying)
                .mapToObj(effortCount -> attemptFetch(uri, effortCount))
                .peek(last::set)
                .filter(Try::isSuccess)
                .findFirst()
                .orElse(last.get());
    }

    private Try<HttpResponse<String>> attemptFetch(URI uri, int effortCount) {
        Try<HttpResponse<String>> newEffort = attempt(() -> createAndSendHttpRequest(uri).get());
        if (newEffort.isFailure()) {
            logger.warn(String.format("Failed HttpRequest on attempt %d of 3: ", effortCount + 1)
                    + newEffort.getException().getMessage(), newEffort.getException()
            );
        }
        return newEffort;
    }

    private CompletableFuture<HttpResponse<String>> createAndSendHttpRequest(URI uri) {
        HttpRequest httpRequest = HttpRequest.newBuilder()
            .GET()
            .header(ACCEPT, APPLICATION_JSON.getMimeType())
            .header(USER_AGENT, NVA_INSTITUTIONS_LIST_CRAWLER)
            .uri(uri)
            .build();
        return httpClient.sendAsync(httpRequest, BodyHandlers.ofString());
    }

    private int waitBeforeRetrying(int effortCount) {
        if (effortCount > FIRST_EFFORT) {
            try {
                Thread.sleep(WAITING_TIME);
            } catch (InterruptedException e) {
                logger.error(LOG_INTERRUPTION);
                throw new RuntimeException(e);
            }
        }
        return effortCount;
    }
```

Comment (brinxmat) [reply src/main/java/no/unit/nva/institution/proxy/HttpExecutorImpl.java:94]: I guess this looks clearer, and I *think* it does what you intend

**Review by brinxmat — DISMISSED** [review]

**Review by brinxmat — APPROVED** [review]

**Pull request merged**